File: q20_Valid_Parentheses/Valid_Parentheses.rb
# @param {String} s
# @return {Boolean}
def is_valid(s)
  result = []
  (0...s.length).each do |i|
    case s[i]
    when '[' then result.push s[i]
    when '{' then result.push s[i]
    when '(' then result.push s[i]
    when '}' then return false if result.empty? || result.pop != '{'
    when ']' then return false if result.empty? || result.pop != '['
    when ')' then return false if result.empty? || result.pop != '('
    end
  end
  result.empty?
end

File: 0020_Valid_Parentheses/Valid_Parentheses.py
# -*- coding: utf-8 -*-

"""
Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

An input string is valid if:

1. Open brackets must be closed by the same type of brackets.
2. Open brackets must be closed in the correct order.

Note that an empty string is also considered valid.

Example 1:

Input: "()"
Output: true

Example 2:

Input: "()[]{}"
Output: true

Example 3:

Input: "(]"
Output: false

Example 4:

Input: "([)]"
Output: false

Example 5:

Input: "{[]}"
Output: true
"""


class Solution:
    def isValid(self, s: str) -> bool:
        result, paren_map = [], {')': '(', ']': '[', '}': '{'}
        for c in s:
            if c not in paren_map:
                result.append(c)
            elif not result or result.pop() != paren_map[c]:
                return False
        return not result
